src/gebt_poc/gen_alpha_2D.cpp

```diff
@@ -192,75 +201,58 @@ std::tuple<State, Kokkos::View<double*>> GeneralizedAlphaTimeIntegrator::AlphaSt
+        auto solution_residual = Kokkos::subview(residuals, Kokkos::make_pair(0ul, size_dofs));
+        KokkosBlas::scal(solution_residual, preconditioning_factor, solution_residual);
+        openturbine::gebt_poc::solve_linear_system(iteration_matrix, residuals);
+        KokkosBlas::scal(residuals, -1, residuals);
 
-            Kokkos::parallel_for(
-                size_dofs,
-                KOKKOS_LAMBDA(const size_t i) { residuals(i) = residuals(i) * kBetaLocal * h * h; }
-            );
-        }
-
-        Kokkos::deep_copy(soln_increments, residuals);
-        openturbine::gebt_poc::solve_linear_system(iteration_matrix, soln_increments);
+        // Update constraints based on the increments
+        auto delta_lagrange_mults =
+            Kokkos::subview(residuals, Kokkos::make_pair(size_dofs, size_problem));
+        KokkosBlas::axpy(1. / preconditioning_factor, delta_lagrange_mults, lagrange_mults_next);
 
+        // Update states based on the increments
         Kokkos::parallel_for(
-            delta_gen_coords.size(),
-            // Take negative of the solution increments to update generalized coordinates
-            KOKKOS_LAMBDA(const size_t i) { delta_x(i) = -soln_increments(i); }
-        );
-
-        if (n_constraints > 0) {
-            if (this->is_preconditioned_) {
+            node_team_policy,
+            KOKKOS_LAMBDA(const member_type& member) {
+                auto node = member.league_rank();
+                constexpr auto components = kNumberOfLieAlgebraComponents;
                 Kokkos::parallel_for(
-                    n_constraints,
-                    KOKKOS_LAMBDA(const size_t i) {
-                        // Take negative of the solution increments to update Lagrange
-                        // multipliers
-                        delta_lagrange_mults(i) =
-                            -soln_increments(i + delta_gen_coords.size()) / (kBetaLocal * h * h);
-                        lagrange_mults_next(i) += delta_lagrange_mults(i);
+                    Kokkos::TeamVectorRange(member, components),
+                    [=](std::size_t i) {
+                        delta_gen_coords(node, i) += residuals(node * components + i);
                     }
                 );
-            } else {
                 Kokkos::parallel_for(
-                    n_constraints,
-                    KOKKOS_LAMBDA(const size_t i) {
-                        // Take negative of the solution increments to update Lagrange
-                        // multipliers
-                        delta_lagrange_mults(i) = -soln_increments(i + delta_gen_coords.size());
-                        lagrange_mults_next(i) += delta_lagrange_mults(i);
+                    Kokkos::TeamVectorRange(member, components),
+                    [=](std::size_t i) {
+                        velocity_next(node, i) += kGammaPrime * residuals(node * components + i);
+                    }
+                );
```

Comment:
why not combine these 3 inner parallel_fors into one?

Reply:
TeamVectorRange should be trying to vectorize the loop under the hood (at least on CPU).  I tried to keep these loops to simple statements to give the compilier a decent chance of doing so successfully.  It also might give the CUDA warps a more regular memory access pattern... I don't have any proof that either of those statements actually make a difference in the final assembly, but they feel right.  Compiler explorer could possibly help here.